[extension/ecsobserver] Write discovered targets as Prometheus file sd

pingleig opened the pull request:
**Description:**

- Glue all pieces using `serviceDiscovery` and support real AWS client
- Write output in prometheus file sd format
- Filter ECS service based on config to reduce API calls

**Link to tracking Issue:**

- Feature request #1395 

Previous PRs

- Fetch EC2, Service https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3503
- Service matcher https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3386
- Exporter https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3333
- Fetch Task https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3284
- Docker Label Matcher https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3276
- Task IP and Port https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3133
- Target and Task definition https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2939
- Config PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377
- Original PR #2734 

**Testing:** 

- unit test
- e2e test using https://github.com/pingleig/aws-otel-collector/commit/d30c480e45073ec72dff51612a884d475d899b81

**Documentation:** 

No new doc. See [existing doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md)

pingleig committed:
ext: ecsobserver Add filter and export yaml

pingleig committed:
ext: ecsobserver Add unit test for overall discovery

- Update README

pingleig committed:
ext: ecsobserver Rename exported structs

pingleig commented:
cc @anuraaga @Aneurysm9

pingleig commented:
Load test failed because some UDP message is lost https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3785/checks?check_run_id=2822779561

```
2021/06/14 18:45:44 Agent process stopped, exit code=0
    validator.go:42: 
        	Error Trace:	validator.go:42
        	            				test_case.go:279
        	            				scenarios.go:190
        	            				log_test.go:177
        	Error:      	Not equal: 
        	            	expected: 139900
        	            	actual  : 139867
        	Test:       	TestLog10kDPS/udp
        	Messages:   	Received and sent counters do not match.
```

mxiamxia commented on extension/observer/ecsobserver/factory.go:
```diff
@@ -42,7 +45,13 @@ func createDefaultConfig() config.Extension {
 	sdCfg := cfg.(*Config)
-	sd, err := NewDiscovery(*sdCfg, ServiceDiscoveryOptions{Logger: params.Logger})
+	opt := serviceDiscoveryOptions{Logger: params.Logger}
+	// Only for test
+	fetcher := ctx.Value(ctxFetcherOverrideKey)
+	if fetcher != nil {
+		opt.FetcherOverride = fetcher.(*taskFetcher)
+	}
+	sd, err := newDiscovery(*sdCfg, opt)
```
If it is for test purpose only, can we move this piece of code in unit test only instead of having it here ?

pingleig replied:
That's for testing create extension, I could pass the ctx to `newDiscovery` and move the logic there. But still it's not test only and always get executed.

Aneurysm9 replied:
I agree with Min that this doesn't feel right.  I'm wary of the amount of "test only" code.  If the fetcher is a dependency then allow dependency injection through configuration or initializer parameters or something, and have tests provide a mock dependency during testing.  Having special cases doesn't seem like the right way to do this.

pingleig replied:
Moved to `Config` as there are just three parameters for creating extension `func createExtension(ctx context.Context, params component.ExtensionCreateSettings, cfg config.Extension)`, originally I don't want to update `Config` as it's more a POJO like struct, but it's a bit type safer than using context and that Config struct is not expected to be used outside package. Though I don't remember reflection based Unmarshal works with unexported struct.

A second thought, I could just have an extra `createExtensionWithFetcher` for test and avoid updating `Config`.

pingleig replied:
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3785/commits/87f107b15e4b3a2d1efd9e3b68c648d58191ac2a

mxiamxia commented on extension/observer/ecsobserver/fetcher.go:
```diff
@@ -86,24 +88,42 @@ func newTaskFetcher(opts taskFetcherOptions) (*taskFetcher, error) {
+	// Match all the services for test. For production if there is no service related config,
+	// we don't describe any service, see serviceConfigsToFilter for detail.
+	if fetcher.serviceNameFilter == nil {
+		fetcher.serviceNameFilter = func(name string) bool {
 			return true
-		},
+		}
 	}
```
Also, can we override `fetcher.serviceNameFilter` with this aways returning `true` function in unit test code?

pingleig replied:
Yeah we can, I was too lazy to update all the test usages, I will create a factory func for tests.

pingleig replied:
Updated in https://github.com/pingleig/opentelemetry-collector-contrib/commit/75c4a9c2becb8baf93ae2b03c94aed944585e56b

mxiamxia commented on extension/observer/ecsobserver/sd.go:
```diff
@@ -17,41 +17,104 @@ package ecsobserver
+			}
+			// We may get 0 targets form some recoverable errors
+			// e.g. throttled, in that case we keep existing exported file.
+			if len(targets) == 0 {
```
do you think it's better to add a log entry here?

pingleig replied:
We can add a debug/warning log here, most errors are already printed in the error handling above. Though it's possible user wrote a config that does not match any tasks.

pingleig replied:
Updated in https://github.com/pingleig/opentelemetry-collector-contrib/commit/75c4a9c2becb8baf93ae2b03c94aed944585e56b

pingleig committed:
ext: ecsobserver Merge duplicated create test fetcher

- Stop the collector process from extension using `host.ReportFatalError`
  otherwiese the failure of extension just log.

mxiamxia reviewed (approved):
Thanks!

Aneurysm9 commented on extension/observer/ecsobserver/sd.go:
```diff
@@ -17,41 +17,112 @@ package ecsobserver
+	if err != nil {
+		return nil, fmt.Errorf("init serivce name filter failed: %w", err)
+	}
+	var fetcher *taskFetcher
+	if opts.FetcherOverride != nil {
+		fetcher = opts.FetcherOverride
+	} else {
+		fetcher, err = newTaskFetcher(taskFetcherOptions{
+			Logger:            opts.Logger,
+			Region:            cfg.ClusterRegion,
+			Cluster:           cfg.ClusterName,
+			serviceNameFilter: svcNameFilter,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init fetcher failed: %w", err)
+		}
+	}
```
This should probably be done by the caller and the `fetcher` provided directly here.  Again, this smells of incomplete dependency injection attempting to support tests, but not fully considered.

pingleig replied:
changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3785/commits/392a893222c9e17750af80680a3f178447c5ab81

Aneurysm9 commented on extension/observer/ecsobserver/target.go:
```diff
@@ -127,3 +140,51 @@ var (
+				delete(labels, k)
+			}
+		}
+		// Rename job label as a workaround for https://github.com/open-telemetry/opentelemetry-collector/issues/575
```
Can you provide more detail on what this is doing and why?  It looks like the referenced issue has been resolved.

pingleig replied:
Actually that issue is two issues, one is dropping job label, another is providing job label in sd or relabel it. I think the comment in that issue explains it pretty well https://github.com/open-telemetry/opentelemetry-collector/issues/575#issuecomment-814558584

pingleig committed:
ext: ecsobserver Explain the rename job lable logic

prom receiver is expecting metric's job name same as the one in config.
In order to keep similar behaviour as cloudwatch agent's discovery impl,
we support getting job name from docker label, but it will break metric
type.

For long term solution, see https://github.com/open-telemetry/opentelemetry-collector/issues/575#issuecomment-814558584

pingleig committed:
ext: ecsoberver Inject test fetcher using config

Was using context.

pingleig committed:
ext: ecsobserver Test fatal error on component.Host

pingleig committed:
ext: ecsobserver Move fetcher injection to its own func

pingleig commented:
windows test failed https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/15444/workflows/90c5815a-0b99-4ab5-ab15-d93e2aeb0575/jobs/134087

```text
--- FAIL: Test_statsdreceiver_EndToEnd (0.00s)
    --- FAIL: Test_statsdreceiver_EndToEnd/default_config_with_9s_interval (0.00s)
        receiver_test.go:132: 
            	Error Trace:	receiver_test.go:132
            	Error:      	Received unexpected error:
            	            	listen udp 127.0.0.1:50835: bind: An attempt was made to access a socket in a way forbidden by its access permissions.
            	Test:       	Test_statsdreceiver_EndToEnd/default_config_with_9s_interval
```

alolita commented:
@dependabot build

pingleig commented:
Did a rebase, unit test passed but load test failed https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/15652/workflows/b630b1c3-7505-47a1-aade-878eb52a8f2a/jobs/136105

```
Failed
2021/06/22 23:11:09 Starting mock backend...
2021/06/22 23:11:09 Starting Agent (/home/circleci/project/bin/otelcontribcol_linux_amd64)
2021/06/22 23:11:09 Writing Agent log to /home/circleci/project/testbed/tests/results/TestLog10kDPS/OTLP/agent.log
2021/06/22 23:11:09 Agent running, pid=34225
2021/06/22 23:11:10 Starting load generator at 10000 items/sec.
2021/06/22 23:11:12 Agent RAM (RES):   0 MiB, CPU: 0.0% | Sent:     17109 items | Received:    16,100 items (5,364/sec)
2021/06/22 23:11:12 Performance error: CPU consumption is 37.3%, max expected is 35%
2021/06/22 23:11:12 Gracefully terminating Agent pid=34225, sending SIGTEM...
2021/06/22 23:11:12 Cannot send logs: failed to push log data via OTLP exporter: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:37133: connect: connection refused"
2021/06/22 23:11:12 Agent process stopped, exit code=0
2021/06/22 23:11:12 CPU consumption is 37.3%, max expected is 35%
    test_case.go:314: CPU consumption is 37.3%, max expected is 35%
2021/06/22 23:11:12 Stopped generator. Sent:     18500 items
2021/06/22 23:11:12 Stopping mock backend...
2021/06/22 23:11:12 Stopped backend. Received:    17,400 items (5,556/sec)
```

btw: it seems github action is still running, so we are testing on both github action and circle CI?

pingleig committed:
ext: ecsobserver Add comment to example config in README

bogdandrutu commented:
@alolita would be good to have a small design about this, not sure why we decided to write this in prometheus format.

pingleig commented:
@bogdandrutu it's mentioned in README https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md#notify-prometheus-receiver-of-discovered-targets Main reason is if we use receiver creator to create new receiver (i.e. pass discovered targets in process and create a new one for each target) the performance is not good  https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1395#issuecomment-747715986 . Eventually it can switch away from writing file in file sd format and support other receivers (e.g. a batch mode to reconfigure receiver in receiver creator framework).

bogdandrutu commented:
/cc @jrcamp I think the performance problem was discuss, is this something we should fix?

mxiamxia commented:
Hi @bogdandrutu , @jrcamp , We had discussed this performance issue before and @kohrapha has explored a few options including implemented a simple `native` Prom receiver as Jay proposed. But writing ECS scrape targets in static file was still the solution we decided to go to unblock AWS to support ECS auto-discovery in OTel.

This PR is the last PR we need to merge to enable ECS Prom metrics auto-discover and we'll launch this feature as Preview in AWS next month. Could you please help to merge this PR so we can be  unblocked?  We'll continue to work on the possible better solution and fixes. 

cc/ @alolita

jrcamp commented:
Yeah I'm going to look into the performance issues of many scrapers with receivercreator. I think this is a decent workaround in the meantime. Will look into it in #3977.

alolita commented:
@bogdandrutu i think it would be useful to do a design on alternatives w pros and cons esp re: perf issues. 

@mxiamxia can you and the team put together a short design proposal outlining the top alternatives.

mxiamxia commented:
> @bogdandrutu i think it would be useful to do a design on alternatives w pros and cons esp re: perf issues.
> 
> @mxiamxia can you and the team put together a short design proposal outlining the top alternatives.

Hi @Alolita, The current implementation is the top alternative for us from the previous discussion. And we have the high level design in [README](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md#notify-prometheus-receiver-of-discovered-targets). We need to have this PR merged ASAP.

alolita commented:
I've marked this PR ready-to-merge. @bogdandrutu please take a look and flag any other concerns you may have.

pingleig commented:
Circle CI unit test failed https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/16013/workflows/41d3192a-5bfb-4867-b721-7d46f891d1da/jobs/139341

```
running go unit test ./... + coverage in /home/circleci/project/exporter/sumologicexporter
--- FAIL: TestPushFailedBatch (21.62s)
    exporter_test.go:292: 
        	Error Trace:	exporter_test.go:292
        	Error:      	Error message not equal:
        	            	expected: "error during sending data: 500 Internal Server Error"
        	            	actual  : "Post \"http://127.0.0.1:33123\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
        	Test:       	TestPushFailedBatch
FAIL
coverage: 95.0% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/sumologicexporter	25.675s
```
